### PR TITLE
ath79: fixes for Senao devices

### PR DESCRIPTION
**Pull request by mcprat**
### ath79: lzma-loader: allow setting custom magic

and allow setting FLASH_MAX

The mtdsplit parser was recently refactored
to allow the kernel to have a custom image header magic

Let's also do this for the lzma-loader

For example:
When implemented together,
this allows the kernel to "appear" to be a rootfs
in order to trick OEM software to write an image
that is actually kernel + rootfs

At the same time, it would boot to openwrt normally
by setting the same magic in DTS

---

### ath79: remove 'fakeroot' for Senao devices

by using a custom kernel header magic
in both OKLI, DTS, and makefile
this hack is not necessary anymore

however, "rootfs" size and checksum
must now be supplied by the factory.bin
because OEM script assumes a squashfs header
and takes size value from the header that would be there,
but this header no longer exists

This allows freeing up of 1 eraseblock
previously used by the "fakeroot" partition
for bypassing the OEM image verification.

Also:
these Senao devices with a 'failsafe' partition
and the tar-gz factory.bin platform would otherwise require
flashing the new upcoming tar-gz sysupgrade.bin afterward.

so this also prevents having to flash twice from OEM

---

### ath79: rename 'engenius' makefile definitions to 'senao'

These recipes and definitions can apply
to devices from other vendors
with PCB boards or SDK produced by Senao
not only the brand Engenius

for example:
Extreme Networks, WatchGuard, OpenMesh,
Fortinet, ALLNET, OCEDO, etc.

so rename all of these items
and move DEVICE_VENDOR from common to generic/tiny.mk

---

### ath79: add factory.bin for ALLNET ALL-WAP02860AC

This device is a Senao-based product
with the tar-gz platform for factory.bin
and checksum verification at boot time
using checksums stored in uboot environment.

Extremely similar hardware/software to Engenius EAP1200H

---

### ath79: cleanup DTS for ALLNET ALL-WAP02860AC

edit nodes to match similar boards
 - eth0
 - pcie0

bump SPI frequency to 40 MHz

remove &pll node: the property is defined already
in qca955x.dtsi

spelling of qca,no-eeprom;
and move to the correct node pcie0

---

### ath79: add Senao 'failsafe' upgrade procedure

Use a similar method for sysupgrade
for Senao boards with a "failsafe" partition
from the OEM factory.bin upgrade method

Make loader partition writable so that sysupgrade works for it

Benefits are:

This allows freeing up of 1 eraseblock
previously used by the "fakeroot" partition
for bypassing the OEM image verification.

These boards are now unbrickable.
Previously, if the real kernel was damaged or missing
the OKLI kernel loader is programmed to halt
and the only way to recover was with UART serial console
because the the bootloader did not check the real kernel.

Consequences are:
 - The user must perform a sysupgrade after factory.bin flash

tested with
ENS202EXT v1

---

### ath79: adjust tiny Engenius APs to 4k blocksize

ath79/tiny kernel config has
CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y
from commit
05d3540

Because of this, these changes are required for 2 reasons:

1.

the boards in ath79/tiny using the new sysupgrade procedure
for Senao devices with a 'failsafe' partition
are failing to reboot to openwrt after a sysupgrade.

the stored checksum is calculated using the length
of the image to be flashed,
and the actual checksum changes due to JFFS2 space
being formatted within the calculated length of the rootfs

2.

for boards with flash chips that support 4k sectors:
saving configuration over sysupgrade is not possible
because sysupgrade.tgz is appended at a 64k boundary
and the mtd parser starts JFFS2 at a 4k boundary.

for boards with flash chips that do not support 4k sectors:
partitioning with 4k boundaries causes a boot loop
from the mtd parser not finding kernel and rootfs.

example:
0x440000 length of kernel + rootfs (from sysupgrade.bin)
0x439000 offset of rootfs_data (from kernel log)

therefore partitioning must be different for these devices
depending on the flash chip model...
so the partitions node is split between dtsi and dts

---

### CODE REVIEW

**Comment by mpratt14** [comment]
the DTSI patches were build tested, I'm in the process of flashing on those boards
for qca955x senao DTSI, &partitions node is needed because some models to be added in the future have 32 MB flash with ART at the end of flash

**Comment by mpratt14** [comment]
@tmn505 are you available to test changes for ALLNET ALL-WAP02860AC ?

File: target/linux/ath79/dts/ar9341_engenius_eap300-v2.dts [hunk 1]
```diff
@@ -1,22 +1,10 @@
 // SPDX-License-Identifier: GPL-2.0-or-later OR MIT
 
-#include "ar9341.dtsi"
-
-#include <dt-bindings/gpio/gpio.h>
-#include <dt-bindings/input/input.h>
-#include <dt-bindings/mtd/partitions/uimage.h>
+#include "ar934x_senao_ap.dtsi"
```
Comment (adschm) [inline target/linux/ath79/dts/ar9341_engenius_eap300-v2.dts:3]: So, you just replace ar9341.dtsi with ar9344.dtsi here? Does not sound right to me ...

Comment (adschm) [reply target/linux/ath79/dts/ar9341_engenius_eap300-v2.dts:3]: I have a better idea, just wait a moment ...

Comment (mpratt14) [reply target/linux/ath79/dts/ar9341_engenius_eap300-v2.dts:3]: very good catch as always Adrian 👍🏼 

it is only qca955x that is common there...I'll fix this right now

Comment (mpratt14) [reply target/linux/ath79/dts/ar9341_engenius_eap300-v2.dts:3]: I was going to move the `ar934x.dtsi` lines back to the DTS of each device

if you have a better idea I'll wait

**Comment by adschm** [comment]
> These recipes and definitions can apply
> to devices from other vendors
> with PCB boards or SDK produced by Senao
> not only the brand Engenius

But is this really about the boards or the SDK and not just about the implementation Engenius uses for it's firmware?

**Comment by adschm** [comment]
The proper way to do the DTSI consolidation would probably be like this:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/senaodtsi

Note that I'm not really convinced yet whether this should be "Senao" or "Engenius". It's even more messy with qca955x where we have Engenius devices with other setup, so the generic name won't be accurate.

This should see some additional eyes and discussion.

File: target/linux/ath79/dts/qca9557_engenius_eap1200h.dts [hunk 2]
```diff
@@ -1,34 +1,11 @@
...
 	compatible = "engenius,eap1200h", "qca,qca9557";
 	model = "EnGenius EAP1200H";
 
-	aliases {
-		label-mac-device = &eth0;
-		led-boot = &led_power;
-		led-failsafe = &led_power;
-		led-running = &led_power;
-		led-upgrade = &led_power;
-	};
```
Comment (adschm) [inline target/linux/ath79/dts/qca9557_engenius_eap1200h.dts:19]: This should be kept in DTS, as leds are defined there.

File: target/linux/ath79/dts/qca9557_engenius_eap1200h.dts [hunk 3]
```diff
@@ -151,24 +53,7 @@
...
 &eth0 {
-	status = "okay";
-
-	mtd-mac-address = <&art 0x0>;
-
 	phy-handle = <&phy5>;
-	phy-mode = "rgmii-id";
-
-	pll-data = <0x82000000 0x80000101 0x80001313>;
-};
```
Comment (adschm) [inline target/linux/ath79/dts/qca9557_engenius_eap1200h.dts:169]: Don't split eth0 across DTS/DTSI. I prefer to keep it in the DTS.

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts [hunk 4]
```diff
@@ -95,88 +56,10 @@
...
-
-	wifi@0,0 {
-		compatible = "qcom,ath10k";
-		reg = <0 0 0 0 0>;
-	};
-};
-
-&pll {
-	clocks = <&extosc>;
-};
```
Comment (adschm) [inline target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:89]: Why is this thrown away?

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts [hunk 5]
```diff
@@ -95,88 +56,10 @@
...
-	status = "okay";
-};
-
 &wmac {
 	status = "okay";
 
 	mtd-cal-data = <&art 0x1000>;
 	mtd-mac-address = <&art 0x0>;
 	mtd-mac-address-increment = <1>;
-	qca-no-eeprom;
```
Comment (adschm) [inline target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:152]: Why is this deleted?

File: target/linux/ath79/dts/qca955x_senao_ap.dtsi [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+				label = "fwconcat0";
+				reg = <0x230000 0xb40000>;
+			};
+
+			partition@d70000 {
+				label = "failsafe";
+				reg = <0xd70000 0x280000>;
+				read-only;
+			};
+		};
```
Comment (adschm) [inline target/linux/ath79/dts/qca955x_senao_ap.dtsi:119]: I'd keep art here until it actually needs to be moved away.

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts (hunk 4 above)
Comment (mpratt14) [reply target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:89]: this is already defined in qca955x.dtsi

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts (hunk 5 above)
Comment (mpratt14) [reply target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:152]: 1. spelled wrong so there is no effect
2. mtd-cal-data takes care of this

File: target/linux/ath79/dts/qca955x_senao_ap.dtsi (hunk 6 above)
Comment (mpratt14) [reply target/linux/ath79/dts/qca955x_senao_ap.dtsi:119]: I have been working on a board for someone that requires it and I will be adding PR for that board right after this

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts (hunk 4 above)
Comment (adschm) [reply target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:89]: Then please address this in a separate commit (a comment would have helped ...)

**Comment by mpratt14** [comment]
> But is this really about the boards or the SDK and not just about the implementation Engenius uses for it's firmware?

We have 3 examples already of other vendors that have almost the same situation (board / SDK from Senao)
 - ALLNET board in this PR
 - OpenMesh products (now that I think of it...they can use this new DTSI)
 - Plasma Cloud products

for OpenMesh dualboot, images are verified almost the same way, and uboot environment has the same variables, plus the extra one for active / inactive boot partition.

> qca955x where we have Engenius devices with other setup

this is true, however for the Engenius devices that do not apply to this, there is `senao-header` command instead which is consistent with `senao-tar-gz`

**Comment by mpratt14** [comment]
@adschm how about a more specific name for DTSI?

`ar724x_senao_failsafe_ap.dtsi`
etc.

File: target/linux/ath79/dts/qca955x_senao_ap.dtsi (hunk 6 above)
Comment (mpratt14) [reply target/linux/ath79/dts/qca955x_senao_ap.dtsi:88]: @adschm  Should I include the new board in this PR or wait until this is done?

Comment (adschm) [reply target/linux/ath79/dts/qca955x_senao_ap.dtsi:88]: You already have at least three different subjects in this PR already, I don't think you will speed it up by adding another one ...

Comment (mpratt14) [reply target/linux/ath79/dts/qca955x_senao_ap.dtsi:88]: BTW you can follow progress on that board here if you like

https://forum.openwrt.org/t/adding-openwrt-support-for-senao-like-devices/88170/

Comment (mpratt14) [reply target/linux/ath79/dts/qca955x_senao_ap.dtsi:88]: @adschm do you still want the DTSI to have ART for now? I would prefer not to have to refactor the DTSI again....

**Comment by ecsv** [comment]
@mpratt14 See https://github.com/ecsv/openwrt/commit/f7c9b4336c05a27af6bca4b635d98de0f886c7c3

**Comment by mpratt14** [comment]
@ecsv ok nice

however I wonder...do you actually need dd for writing? for my case, it is only needed for the checksum

Also, I had significant trouble with doing mtd erase first and write later so I took the model from default_do_upgrade instead
the other way caused some blocks at the beginning of the partition to be lost when appending the backup file to JFFS2

mtd has an option for offset I believe would work as well

`-p <number>             write beginning at partition offset`

**Comment by ecsv** [comment]
Since the data has to be extracted from the CE01 image - yes, dd is needed (or some other tool to extract the data from the correct position). And if you're data is correctly aligned with the erase block size then you should not have problems with "lost blocks"

And the script is already using -p for mtd.

**Comment by mpratt14** [comment]
yeah I cannot explain why I was having that issue...the only difference is using mtd as erasing and writing at the same time, instead of separately...and only with the backup file

**Comment by mpratt14** [comment]
@adschm I discovered a small regression in the tiny subtarget that has been giving me problems...

4k mtd erasesize was enabled by this commit
05d35403b211ccb9bf20e2b3c0b7a30c82c6d6a2

every mtd device is showing up as 4k erasesize
however in the makefiles, all of ath79 is set to 64k
`BLOCKSIZE := 64k`

this is causing rootfs to be padded to 64k before the 0xdeadcode marker for JFFS2
then the mtd parser sees empty space before the marker and starts JFFS2 there instead

for example in my case, kernel + rootfs length is `0x440000` + 4 bytes for 0xdeadcode
on the next boot, rootfs_data partition shows as starting at `0x439000`
after another reboot, the checksum fails since `0x439000` to `0x440000` was changed

this breaks the checksum that I'm trying to set in the uboot environment
and also, saving the configuration during sysupgrade is not possible...

I am guessing the best solution is to set all of tiny subtarget to `BLOCKSIZE := 4k`
but I am not sure the best way to do this

**Comment by mpratt14** [comment]
nevermind...it has to be handled per device...

**Comment by mpratt14** [comment]
@adschm ready for review again

refactored the sysupgrade script to save memory

added blocksize fix for the boards in tiny

moved the changes to ALLNET to the bottom so they are easier to drop
(in case they never get tested)

tested on all boards except EnstationAC

**Comment by mpratt14** [comment]
updated commit message

**Comment by adschm** [comment]
> ar724x_senao_failsafe_ap.dtsi

Why "failsafe"?

**Comment by adschm** [comment]
I still don't see why a DTSI for an arbitrary choice of devices should get a generic name like qca955x_senao_ap.dtsi, while other "Senao AP" will have different DTSes.

**Comment by adschm** [comment]
What about
qca955x_senao_okli.dtsi
qca955x_senao_loader.dtsi
qca955x_senao_concat.dtsi
...?

**Comment by mpratt14** [comment]
I was thinking "failsafe" because they all have a "failsafe" partition with a linux image for uploading OEM or factory.bin
the upgrade page looks like this
https://forum.openwrt.org/uploads/default/original/2X/3/3348c466c9625e47230dd32f5669debd4d0bbcf4.jpeg

"loader" is probably better though

**Comment by mpratt14** [comment]
@adschm ok DTSI are now "senao_loader"

**Comment by adschm** [comment]
I will push a patch for the qca955x DTSI soon as well, so maybe rebase _after_ that.

Note that since I've merged the DTSIs but not the failsafe patch now, you will have to adjust that.

Both changes won't probably be quick, and I doubt I can review either of them properly.
Any testers for ALLNET yet?

**Comment by tmn505** [comment]
How am I supposed to use this factory.bin image? When I uploaded the image in web interface and proceeded with upgrade, it looked as if the upgrade went ok, but after reboot the system is still from OEM. Different behavior was observed when I used failsafe. The upgrade was concluded successful after 1s and the U-Boot variables regarding checksums were wiped. The original system is still on the flash and device is stuck in failsafe mode.
What reference did You use when creating the factory.bin image? Did You check it with available source code (https://www.allnet.de/ftp-downloads/allnet/wireless/all-wap0305ac/Sourcecode.zip)?

**Comment by mpratt14** [comment]
@tmn505 Hi, sorry for the confusion.

The factory.bin should have the same result with both OEM upgrade page and with the failsafe image, it's interesting to hear otherwise but we will figure it out...

Their website does not offer firmware download for this model, otherwise, I would have grabbed that image and unpacked it to see for myself...
The only reference right now is the partitions (the fact that there is a failsafe image at all) and many other similarities in the DTS to Engenius devices I have worked with. Most likely, there is a simple change or addition we need to make for ALLNET devices

If my assumptions are correct, there would be a file
`/etc/fwupgrade.sh`
in both the OEM firmware and in the failsafe image (accessible if you have root shell on either one)

That script should be all we need in order to make the adjustments necessary for the factory.bin to work

If you have a hard time getting root access, we can also isolate the rootfs from a flash dump and unsquash it

In the meantime, I will see if this file is in their source code package.
Also, I understand now that there was a custom ALLNET sysupgrade in ar71xx, I will have to take a look at that too...

**Comment by mpratt14** [comment]
> variables regarding checksums were wiped

u-boot will automatically delete those variables when verifying the checksums fail

probably what happened is, the new variables were written, but the images were not

**Comment by mpratt14** [comment]
@tmn505 I did find the upgrade script in the ALLNET GPL package, it looks the same as those from Engenius

./custom/feeds/senaoecm/ecm-web/failsafe/senao_v2/common/root/etc/fwupgrade.sh
./custom/feeds/senaoecm/ecm-web/target/senao_v2/common/root/etc/fwupgrade.sh

It might be that the only thing incorrect is the "modelname"

the script's "errcode" variable is bypassed, so part of the script can fail without triggering a return and telling the user that something was wrong

if you could give the output of
`cat /etc/modelname`

**Comment by mpratt14** [comment]
I am testing one more idea...so we can get rid of the fakeroot from factory.bin...

**Comment by tmn505** [comment]
Added rootfs copy for whole picture.
[rootfs.bin.gz](https://github.com/openwrt/openwrt/files/6043374/rootfs.bin.gz)

**Comment by mpratt14** [comment]
@tmn505 thanks, it is ready to try a factory.bin again

**Comment by mpratt14** [comment]
@blocktrron The first commit right now in this PR is a *very small* change to OKLI lzma-loader. This PR is an alternative to my other PR that had large changes to the loader. I hope it is acceptable. If there is a better way to handle the variable, please advise me and I'll make that change. Thanks

File: target/linux/ath79/dts/ar7240_engenius_enh202-v1.dts [hunk 7]
```diff
@@ -71,6 +71,24 @@
 	};
 };
 
+&partitions {
```
Comment (adschm) [inline target/linux/ath79/dts/ar7240_engenius_enh202-v1.dts:74]: I don't think the DTSI should be cut down further here. Since the DTSI only contains the partitioning anyway (and actually all devices using it are 4k), please just apply the changes to the DTSI. When we actually will have multiple partitioning schemes for ar724x at some point, we can revisit this and rename the DTSI to something-8m.dtsi eventually.

Comment (mpratt14) [reply target/linux/ath79/dts/ar7240_engenius_enh202-v1.dts:74]: the mutliple partitioning schemes is actually required now (to fix config restore)

ENH202 v1 is 64k
the other boards are 4k

I will go ahead and make the 2 different DTSIs now

Comment (adschm) [reply target/linux/ath79/dts/ar7240_engenius_enh202-v1.dts:74]: Err, just looked at it properly: Why is enh202-v1 not set to 4k, though it's tiny as well?
Is that what you mean by "therefore partitioning must be different for these devices depending on the flash chip model..."? Please make that clearer, also by actually naming the devices you speak about.
Even if that holds, I'd not split down the DTSI further, and create something-4k.dtsi and something-64k.dtsi instead.

Comment (mpratt14) [reply target/linux/ath79/dts/ar7240_engenius_enh202-v1.dts:74]: Yes exactly. It depends on the flash chip model. I will make this more clear in the commit message

for ENH202, the kernel detects the flash as "m25p64" which does not have SECT_4K symbol

`[    0.332017] spi-nor spi0.0: m25p64 (8192 Kbytes)`

https://github.com/torvalds/linux/blob/master/drivers/mtd/spi-nor/micron-st.c#L197

Comment (mpratt14) [reply target/linux/ath79/dts/ar7240_engenius_enh202-v1.dts]: ok done 👍🏼

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts [hunk 8]
```diff
@@ -75,67 +77,21 @@
...
-			art: partition@ff0000 {
-				label = "art";
-				reg = <0xff0000 0x010000>;
-				read-only;
-			};
-		};
+	wifi@0,0,0 {
+		compatible = "qcom,ath10k";
+		reg = <0x0 0 0 0 0>;
+		qca,no-eeprom;
```
Comment (tmn505) [inline target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:94]: Drop this property, ath10k doesn't understand it.

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts (hunk 5 above)
Comment (tmn505) [reply target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:148]: Ack, because it was mistyped 'mtd-cal-data' actually worked.

**Review by tmn505 — CHANGES_REQUESTED** [review]
Now the image flashing works, with the small change taken care of, it's good to go.

File: target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts (hunk 8 above)
Comment (mpratt14) [reply target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:94]: are you sure? the reason I put this is because you had this in the caldata script which suggests the caldata is on flash instead of eeprom...

specifically the art partition from
0x5000
to
0x5844

```
	allnet,all-wap02860ac|\
	glinet,gl-x750)
		caldata_extract "art" 0x5000 0x844
		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0x0) +2)
		;;
```

what does kernel log say exactly?

also if you are not sure, you can dump the art partition and upload it for us to see
(I think it's mtd7, verify with `cat /proc/mtd`)

`hexdump -C /dev/mtd7`

Comment (tmn505) [reply target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:94]: There's no point in what You're trying discuss here. The point is, even if You add this property here, it will be ignored by the driver. Run `grep -R 'no-eeprom' drivers/net/wireless/ath` in root of Linux source tree, You'll see that only driver which understands this is ath9k.

Comment (mpratt14) [reply target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:94]: I just checked the ath10k docs

Sorry, all this time I did not realize 😆

**Comment by mpratt14** [comment]
@adschm updated commit messages, and removed qca,no-eeprom

and one last commit at the end, if you don't like it you can drop it 👍🏼

**Comment by mpratt14** [comment]
right now we have 

```
  IMAGE/factory.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | pad-rootfs | \
	check-size | senao-tar-gz $$$$(SENAO_IMGNAME)
  IMAGE/sysupgrade.bin := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | \
	pad-rootfs | check-size | senao-tar-gz $$$$(SENAO_IMGNAME) | append-metadata
```

is this acceptable?

```
  IMAGE/sysupgrade.bin := $(IMAGE/factory.bin) | append-metadata
```

**Comment by mpratt14** [comment]
@adschm Hi Adrian, I do not mean to "rush" the process, but I am hoping to get these changes into the new release branch. I believe this is the last set of changes that are needed for complete support that fully handles all cases with these boards.

Also as I said, you can drop the last commit if you want to leave that DTSI file alone.

As always, I appreciate your reviews, and if there is anything that needs to be changed, let me know and I will handle it immediately.

thanks :D

**Comment by mpratt14** [comment]
edited commit title and rebased

**Comment by mpratt14** [comment]
@blocktrron any comment on the first commit? do you like the idea?

**Comment by mpratt14** [comment]
@adschm ping

**Comment by mpratt14** [comment]
@adschm rebased and edited commit messages to be much more clear and concise

**Comment by mpratt14** [comment]
@adschm may I please have a (final?) review

**Comment by mpratt14** [comment]
@ynezz this is ready for merge IMO. I have tested these commits on 10+ boards (some of which are not added yet).

Adrian gave me excellent and tough reviews and thanks to him, this PR is now in great condition. However, he has been silent for a month now (I assume he is busy with other things).

This should be in the stable release because of the added sysupgrade platform, and otherwise, there would be migration for the affected boards between this coming stable release and the next one. Also, this PR adds a factory.bin for a device that would otherwise require UART to flash.

In order for it to safely land in stable release without any issue, let's get it into master now so I can flash snapshots and verify that everything is ok....

thanks for your time

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

**Comment by adschm** [comment]
Thanks for taking care, this was too big for me at the moment.

**Comment by mpratt14** [comment]
@danielfdickinson looking at the old ar71xx sysupgrade platforms (allnet.sh), this Senao platform for factory.bin and sysupgrade should apply to your devices as well (CAP324, CR3000, CR5000)

it should be easy to add those to this....if you can verify that they have the same situation (OEM has a failsafe image thats booted when checksums fail, and do some testing...

these 2 commits are good examples, probably we can do it all in 1 commit though
37ea5d9a658a15783b0ac68583db9e67a43e960d
e800da9d5c87d276806284cf2183994ea7fe083e

**Comment by danielfdickinson** [comment]
@mpratt14 The CAP324, CR3000, and CR5000 had different bootloader and don't have a failsafe image, at least to my knowledge. I will try to make some time to look closer, but I don't think these commits will works with these devices.

**Comment by mpratt14** [comment]
@danielfdickinson if that turns out to be true, I'm curious why it used the same sysupgrade script as these boards would use in the old target ar71xx. Is it possible that theres no failsafe image, but checksums are still verified at boot? and it does something else instead when it fails?

**Comment by danielfdickinson** [comment]
Since I added the ar71xx support, and I don't totally remember, my guess is that it was based on the CAP4200, even though the bootloader is different. I will try and make some time this month, but I'm pretty sure the CAP324 an CR3000/CR5000 do not check even check a checksum, I do know they don't have enough flash for a fallback (8M flash, fully utilized by the stock firmware for the CR3000/CR5000). The CAP324 had 16M flash but didn't split into a fallback.

**Comment by danielfdickinson** [comment]
Ok, so the PowerCloud stuff (CR3000, CR5000, and CAP324) don't have a failsafe. (Sorry for taking so long to get back to you). They used a u-boot that was modified from Senao's default u-boot. From what I can recall the reason for using the the same script many moons ago was that the boards were simply basing the sysupgrade on what worked for similar hardware. It worked, but was  probably not ideal. IIRC they now use a generic sysupgrade (for ath79).

In any event I'm currently not using any of these devices, although I see someone is getting APXXX from Senao working based on the CAP324. Personally I would not take it badly if PowerCloud stuff were removed after the upcoming release, or even for this release, although at this late date there is always the chance there is someone else using it and who would be upset.

**Comment by mpratt14** [comment]
@danielfdickinson ok no problem. my only question would be how the OEM sysupgrade worked in order to maybe have a factory.bin for that device

do you have an original upgrade image maybe? or a flash backup?

**Comment by danielfdickinson** [comment]
Ah, that I can answer. For the CR3000 and CR5000 the sysupgrade was entirely cloud-based -- there was no local firmware upgrade option. In the background it was a standard OpenWrt sysupgrade (the OEM firmware was OpenWrt-based not the usual Senao thing). For the CAP324 the upgrade was _usually_ cloud-based but there was the option of using a local LuCI-based interface to do a sysupgrade as well (again OpenWrt-based firmware).

The details are in the Table of Hardware on the Wiki:

1. For the CAP324: https://openwrt.org/toh/powercloud_systems/cap324
2. For the CR5000: https://openwrt.org/toh/powercloud_systems/cr5000
3. For the CR3000: https://openwrt.org/toh/powercloud_systems/cr3000

HTH.
